### PR TITLE
[feature](function) Modified cast as time to behave more like MySQL 

### PR DESCRIPTION
**Pull request by Mryange**
# Proposed changes
Because the underlying type of time was float64, select cast("19:22:18" as time) would result in a null value in the past.
Results in the following:
**before**

```
mysql [(none)]>select cast("19:22:18" as time);
+--------------------------+
| CAST('19:22:18' AS TIME) |
+--------------------------+
| NULL                     |
+--------------------------+

mysql [(none)]>select cast(300 as time);
+-------------------+
| CAST(300 AS TIME) |
+-------------------+
| 00:05:00          |
+-------------------+

mysql [(none)]>select cast(360 as time);
+-------------------+
| CAST(360 AS TIME) |
+-------------------+
| 00:06:00          |
+-------------------+
```

**affter**
```
mysql [(none)]>select cast("19:22:18" as time);
+--------------------------+
| CAST('19:22:18' AS TIME) |
+--------------------------+
| 19:22:18                 |
+--------------------------+

mysql [(none)]>select cast(300 as time);
+-------------------+
| CAST(300 AS TIME) |
+-------------------+
| 00:03:00          |
+-------------------+

mysql [(none)]>select cast(360 as time);
+-------------------+
| CAST(360 AS TIME) |
+-------------------+
| NULL              |
+-------------------+
```

**Support for some features of MySQL**
```
mysql [(none)]>select cast('300' as time);
+---------------------+
| CAST('300' AS TIME) |
+---------------------+
| 00:03:00            |
+---------------------+

mysql [(none)]>select cast('300:20' as time);
+------------------------+
| CAST('300:20' AS TIME) |
+------------------------+
| 300:20:00              |
+------------------------+
```

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

File: be/src/vec/functions/function_cast.h [hunk 1]
```diff
@@ -1805,6 +1921,10 @@ class FunctionCast final : public IFunctionBase {
 
             return false;
         };
+        if (to_type->is_time()) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_cast.h:1924]: use call_index_id to do the work

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p1

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by HappenLee** [comment]
run buildall

File: be/src/vec/functions/function_cast.h [hunk 2]
```diff
@@ -62,7 +65,91 @@ inline UInt32 extract_to_decimal_scale(const ColumnWithTypeAndName& named_column
...
+            auto int_value = StringParser::string_to_unsigned_int<uint64_t>(
+                    reinterpret_cast<char*>(begin), len, &parse_result);
+            if (UNLIKELY(parse_result != StringParser::PARSE_SUCCESS)) {
+                return false;
+            }
+            num = int_value;
+            return true;
+        };
+        char *first_colon {nullptr}, *second_colon {nullptr};
+        if ((first_colon = (char*)memchr(first_char, ':', len)) != nullptr) {
```
Comment (BiteTheDDDDt) [inline be/src/vec/functions/function_cast.h:91]: better use `If statements with initializer`
if(x=y;x){
}

File: be/src/vec/functions/function_cast.h [hunk 3]
```diff
@@ -62,7 +65,91 @@ inline UInt32 extract_to_decimal_scale(const ColumnWithTypeAndName& named_column
...
+            if ((second_colon = (char*)memchr(first_colon + 1, ':', end_char - first_colon - 1)) !=
+                nullptr) {
+                // find tow colon
```
Comment (BiteTheDDDDt) [inline be/src/vec/functions/function_cast.h:94]: tow->two

File: be/src/vec/functions/function_cast.h [hunk 4]
```diff
@@ -62,7 +65,91 @@ inline UInt32 extract_to_decimal_scale(const ColumnWithTypeAndName& named_column
...
+        return true;
+    }
 
+    // Cast from number
+    template <typename T, typename S>
+    static bool try_parse_time(T from, S& x) {
+        int64 seconds = from;
+        int64 hour = 0, minute = 0, second = 0;
+        second = seconds % 100;
+        seconds /= 100;
```
Comment (BiteTheDDDDt) [inline be/src/vec/functions/function_cast.h:142]:
```
seconds=from/100
second=from-seconds*100
```
Maybe performance will be better this way

File: be/src/vec/functions/function_cast.h [hunk 5]
```diff
@@ -227,11 +314,25 @@ struct ConvertImpl {
                     }
                 }
             } else {
-                for (size_t i = 0; i < size; ++i) {
-                    vec_to[i] = static_cast<ToFieldType>(vec_from[i]);
+                if constexpr (IsDataTypeNumber<FromDataType> &&
+                              std::is_same_v<ToDataType, DataTypeTime>) {
+                    // 300 -> 00:03:00  360 will be parse failed , so value maybe null
+                    ColumnUInt8::MutablePtr col_null_map_to;
+                    ColumnUInt8::Container* vec_null_map_to [[maybe_unused]] = nullptr;
```
Comment (BiteTheDDDDt) [inline be/src/vec/functions/function_cast.h:321]: why have maybe_unused here?

**Comment by Mryange** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run clickbench

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run clickbench

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run clickbench

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**